File: code/backend/src/main/java/com/taskflow/modules/reminder/service/ReminderService.java
package com.taskflow.modules.reminder.service;

import com.taskflow.modules.reminder.dto.CreateReminderRequest;
import com.taskflow.modules.reminder.dto.ReminderDto;
import com.taskflow.modules.reminder.dto.UpdateReminderRequest;

import java.util.List;
import java.util.UUID;

/**
 * Domain Service interface for managing scheduled reminders, status lifecycles, and background event processing.
 */
public interface ReminderService {

    /**
     * Creates a new scheduled reminder for a task.
     *
     * @param userId  UUID identifier of requesting user
     * @param taskId  UUID identifier of target task
     * @param request creation request payload
     * @return ReminderDto instance
     */
    ReminderDto createReminder(UUID userId, UUID taskId, CreateReminderRequest request);

    /**
     * Gets all active reminders for a task.
     *
     * @param userId UUID identifier of requesting user
     * @param taskId UUID identifier of target task
     * @return list of ReminderDto instances
     */
    List<ReminderDto> getTaskReminders(UUID userId, UUID taskId);

    /**
     * Gets user's upcoming pending reminders.
     *
     * @param userId UUID identifier of requesting user
     * @return list of pending ReminderDto instances
     */
    List<ReminderDto> getUserUpcomingReminders(UUID userId);

    /**
     * Gets reminder details by ID.
     *
     * @param userId     UUID identifier of requesting user
     * @param reminderId UUID identifier of target reminder
     * @return ReminderDto instance
     */
    ReminderDto getReminderDetails(UUID userId, UUID reminderId);

    /**
     * Updates reminder scheduled time or status.
     *
     * @param userId     UUID identifier of requesting user
     * @param reminderId UUID identifier of target reminder
     * @param request    update request payload
     * @return updated ReminderDto instance
     */
    ReminderDto updateReminder(UUID userId, UUID reminderId, UpdateReminderRequest request);

    /**
     * Dismisses a triggered or pending reminder.
     *
     * @param userId     UUID identifier of requesting user
     * @param reminderId UUID identifier of target reminder
     * @return updated ReminderDto instance with status DISMISSED
     */
    ReminderDto dismissReminder(UUID userId, UUID reminderId);

    /**
     * Soft-deletes a reminder.
     *
     * @param userId     UUID identifier of requesting user
     * @param reminderId UUID identifier of target reminder
     */
    void deleteReminder(UUID userId, UUID reminderId);

    /**
     * Processes all due pending reminders, updating status to TRIGGERED and publishing domain events.
     * Called by background scheduler.
     */
    void processDueReminders();
}

File: code/frontend/src/features/reminder/services/reminder-service.ts
import { apiClient } from '@/lib/api-client';
import type { ApiResponse } from '@/features/auth/types';
import type { CreateReminderInput, ReminderDto, UpdateReminderInput } from '../types';

export const reminderService = {
  getTaskReminders: async (taskId: string): Promise<ReminderDto[]> => {
    const res = await apiClient.get<ApiResponse<ReminderDto[]>>(`/tasks/${taskId}/reminders`);
    return res.data.data;
  },

  getUserUpcomingReminders: async (): Promise<ReminderDto[]> => {
    const res = await apiClient.get<ApiResponse<ReminderDto[]>>('/reminders/user');
    return res.data.data;
  },

  getReminderDetails: async (reminderId: string): Promise<ReminderDto> => {
    const res = await apiClient.get<ApiResponse<ReminderDto>>(`/reminders/${reminderId}`);
    return res.data.data;
  },

  createReminder: async (taskId: string, data: CreateReminderInput): Promise<ReminderDto> => {
    const res = await apiClient.post<ApiResponse<ReminderDto>>(`/tasks/${taskId}/reminders`, data);
    return res.data.data;
  },

  updateReminder: async (reminderId: string, data: UpdateReminderInput): Promise<ReminderDto> => {
    const res = await apiClient.put<ApiResponse<ReminderDto>>(`/reminders/${reminderId}`, data);
    return res.data.data;
  },

  dismissReminder: async (reminderId: string): Promise<ReminderDto> => {
    const res = await apiClient.patch<ApiResponse<ReminderDto>>(`/reminders/${reminderId}/dismiss`);
    return res.data.data;
  },

  deleteReminder: async (reminderId: string): Promise<void> => {
    await apiClient.delete<ApiResponse<void>>(`/reminders/${reminderId}`);
  },
};
